src/odm_sharing/private/cons.py

```diff
@@ -23,37 +34,186 @@ def _write_table_to_db(db: sa.engine.Engine, table: str, df: pd.DataFrame
+        for val in df[col]:
+            norm_val = val.upper()
+            if norm_val in NA_VALS:  # NA
+                continue
+            if norm_val in BOOL_VALS:  # bool
+                result.add(str(col))
+            break
+    return result
+
+
 def _connect_csv(path: str) -> Connection:
     '''copies file data to in-memory db
 
     :raises OSError:'''
+    # XXX: NA-values are not normalized to avoid mutating user data (#31)
     logging.info('importing csv file')
     table = Path(path).stem
-    db = _create_memory_db()
-    df = pd.read_csv(path)
-    _write_table_to_db(db, table, df)
-    return db
+    df = pd.read_csv(path, na_filter=False)
+    bool_cols = _get_csv_dataset_bool_cols(df)
+
+    # normalize bool (string) values to 0/1
+    #
+    # XXX: Bool columns with NA-values still represent bools as FALSE/TRUE.
+    # This needs to be normalized to 0/1, like normal bool columns, so that all
+    # bool columns have the same bool values.
+    for col in bool_cols:
+        if df[col].dtype == object:  # str
+            df[col] = df[col].replace({'FALSE': '0', 'TRUE': '1'})
+
+    db = _datasets_to_db({table: df})
+    return Connection(db, {table: bool_cols})
+
+
+def _get_excel_bool_col_ixs(data_sheet: Sheet, formula_sheet: Sheet
+                            ) -> List[int]:
+    '''returns the indices of all bool columns'''
+    # XXX: iter_rows gives a type error with mypy
+    result = []
+    header_row = next(data_sheet.iter_rows())  # type: ignore
+    column_count = len(header_row)
+    remaining = list(range(column_count))
+
+    def iter_col_cells(sheet: Sheet) -> Generator:
+        '''iterates column-indices and cells, for each row in the sheet'''
+        # XXX: the read-only workbook only has row iteration
+        # XXX: the first row is skipped since it's assumed to be the header
+        # XXX: iteration of `remaining` must use a copy since it may be mutated
+        for row in sheet.iter_rows(min_row=2):  # type: ignore
+            if len(remaining) == 0:
+                break
+            for col_ix in remaining.copy():
+                cell = row[col_ix]
+                yield (col_ix, cell)
+
+    def add_bool_col(col_ix: int) -> None:
+        result.append(col_ix)
+        remaining.remove(col_ix)
+
+    # find bool cols in data sheet
+    for col_ix, cell in iter_col_cells(data_sheet):
+        if isinstance(cell.value, bool):
+            add_bool_col(col_ix)
+
+    # find bool cols in formula sheet
+    #
+    # XXX: LibreOffice may convert `FALSE`/`TRUE` to `=FALSE()`/`=TRUE()`,
+    # which evaluates to 0/1 which in turn can't be detected as bool, so we
+    # have to check the formulas as well.
+    BOOL_FORMULA_VALS = ['=FALSE()', '=TRUE()']
+    for col_ix, cell in iter_col_cells(formula_sheet):
+        if cell.value in BOOL_FORMULA_VALS:
+            add_bool_col(col_ix)
+
+    return result
+
+
+def _get_excel_bool_cols(path: str, wb: Workbook, included_tables: Set[str]
```

Comment:
Add comment for the path parameter

Reply:
is this still an issue after the last rewrite?